Specify uniqueness of tracestate keys as the sender's responsibility (#477)
w3c/distributed-tracing: spec/20-http_request_header_format.md

### Combined Header Value

The `tracestate` value is the concatenation of trace graph key/value pairs.

Example: `vendorname1=opaqueValue1,vendorname2=opaqueValue2`

> Only one entry per key is allowed. For example, if a vendor name is Congo and a trace started in their system and then went through a system named Rojo and later returned to Congo, the `tracestate` value would not be:

Reviewer: "is allowed" is not a normative language. I would change it to "There SHOULD be only one entry per key".

Author: The normative language around this is in the section "Mutating the tracestate Field":
> Adding a key/value pair MUST NOT result in the same key being present multiple times.

I would restrict the normative description of that aspect to the "Mutating" section because it describes _what an implementation is allowed to do_ with the header and what not. I do not see any added benefit by duplicating this aspect in normative language here. "Only one entry per key is allowed." is given in the context of the example (see the line above) and in my understanding, examples are never normative.

Example: `vendorname1=opaqueValue1,vendorname2=opaqueValue2`

> Only one entry per key is allowed because the entry represents that last position in the trace. Hence vendors must overwrite their entry upon reentry to their tracing system.

Author: "because the entry represents that last position in the trace." had been called out [here](https://github.com/w3c/trace-context/issues/446#issuecomment-759073256) by @yurishkuro as not making a lot of sense, and I agree. Hence, I removed that part.

Also, I think update could be a better phrase instead of overwrite because we also use that in the section "Mutating the traceparent Field". 

Finally "upon reentry to their tracing system" is really unwieldy, IMO.

> Only one entry per key is allowed. Hence vendors MUST update their entry instead of adding it twice when propagating a modfied `tracestate` header downstream.

Reviewer: * the word "hence" not needed
* instead of speaking of updating/adding, I would rather say vendors MUST ensure that the entry with their vendor key is only present once in the header sent downstream.

Reviewer 2: +1 to just say that vendor is responsible for not duplicating keys it owns

Author: Well, that is exactly what I have added down in the "Mutating the traceparent Field" section -> https://github.com/w3c/trace-context/pull/477/files#diff-3832602716774745d030f22592f4c5b7d8f316e4d44be0165282244f9d6f14d9R358

Do we just want to duplicate that phrasing here? Seems redundant. How about doing away with the second sentence entirely? So just "Only one entry per key is allowed." and that's it. We also have the cross-reference 

> See [Mutating the tracestate Field](#mutating-the-tracestate-field) for details. 

so I think spelling out specifics here too is not required, the actual normative rules belong in the "Mutating" section and are already phrased as you requested.

WDYT?

Reviewer 2: Yes, perhaps just delete the entire phrase including "Only one entry per key is allowed."

Reviewer 3: I would keep the first sentence.

Author: I kept the first sentence, it leads nicely into the following example, but I removed the duplication about not adding your own header twice.

Following are allowed mutations:

- **Add a new key/value pair**. The new key/value pair SHOULD be added to the beginning of the list. Adding a key/value pair MUST NOT result in the same key being present multiple times.
- **Update an existing value**. The value for any given key can be updated. Modified keys SHOULD be moved to the beginning (left) of the list.

> - **Delete a key/value pair**. Any key/value pair MAY be deleted. Vendors SHOULD NOT delete keys that were not generated by them. The deletion of an unknown key/value pair will break correlation in other systems. This mutation enables three scenarios. The first is that proxies can block certain `tracestate` keys for privacy and security concerns. The second scenario is a truncation of long `tracestate`s. Finally, vendors MAY also discard duplicate keys that were not generated by them, however, they are not obliged to do that.

Reviewer 4: ```suggestion
- **Delete a key/value pair**. Any key/value pair MAY be deleted. Vendors SHOULD NOT delete keys that were not generated by them. The deletion of an unknown key/value pair will break correlation in other systems. This mutation enables three scenarios. The first is that proxies can block certain `tracestate` keys for privacy and security concerns. The second scenario is a truncation of long `tracestate`s. The third scenario is vendors MAY discard duplicate keys that were not generated by them.
```

Author: "The third scenario is vendors MAY discard" sounds a bit awkward IMO. I would prefer the current phrasing, to be honest.

Reviewer 4: The current phrasing of "Finally, vendors..." is fine with me. However, the phrase ", however, they are not obliged to do that" feels redundant since it is conveyed by the MAY semantics in the previous sentence, so it feels we can remove that part to make it tighter.

Author: Good call, I have removed that part.

Following are allowed mutations:

> - **Add a new key/value pair**. The new key/value pair SHOULD be added to the beginning of the list. Adding a key/value pair MUST NOT result in the same key being present multiple times.

Author: This had been called out in "Combined Header Value" already, but it actually belong here, IMO.

Reviewer 2: I like how this formulated. It eliminates concern that if platform doesn't deal with tracestate keys, just propagates it as is - it is not responsible for the duplicates. We may want to spell it out explicitly

Author: See this part of my PR comment:

> There seems to be no consensus yet whether an implementation should
> clean up a tracestate value it receives by discarding duplicate keys
> that it does not own. Thus, that aspect is not considered in this
> change.

And my message here https://github.com/w3c/trace-context/issues/446#issuecomment-966451599

TL;DR: I agree with you that this is how it should work (vendors should not need to care about foreign keys at all) but I am not sure if this is consensus, so for now I was reluctant to add this aspect to this PR.

Reviewer 3: I agree that a tracing system should be responsible for deduping their own keys and shouldn't need to analyze or dedupe other keys.

Author: I've added that aspect as well now: https://github.com/w3c/trace-context/pull/477/commits/ccbabd6087a3ee9e65889fa54a23f32240436b06
